### PR TITLE
Update linkerd authz command to support authorizationpolicy and httproute

### PR DESCRIPTION
**Pull request by adleong**
Previously, the `linkerd authz` command would list all ServerAuthorization resources which targeted the specified resource.  With the addition of AuthorizationPolicies, we update this command to also show all AuthorizationPolicies which target the specified resource.  In cases where the AuthorizationPolicy targets an HTTPRoute which belongs to the resource, we also print the HTTPRoute name.

Sample output:

```
linkerd authz -n emojivoto po           
ROUTE            SERVER         AUTHORIZATION_POLICY   SERVER_AUTHORIZATION   
*                emoji-grpc     emoji-grpc                                    
linkerd-metrics  linkerd-admin  linkerd-metrics                               
linkerd-probes   linkerd-admin  linkerd-probes                                
*                prom           prom                                          
*                voting-grpc    voting-grpc                                   
*                web-http       web-public
```

Notice that since this command can print a mix of ServerAuthorizations and AuthorizationPolicies, we have a column for each.  Any individual row will only use one or the other of these columns, never both.

### CODE REVIEW

File: pkg/k8s/policy.go [hunk 1]
```diff
@@ -55,47 +52,87 @@ func ServerAuthorizationsForResource(ctx context.Context, k8sAPI *KubernetesAPI,
...
+				fmt.Fprintf(os.Stderr, "ServerAuthorization/%s targets Server/%s but we failed get it: %s\n", saz.Name, saz.Spec.Server.Name, err)
+				continue
 			}
 			servers = []serverv1beta1.Server{*server}
 		} else if saz.Spec.Server.Selector != nil {
 			selector, err := metav1.LabelSelectorAsSelector(saz.Spec.Server.Selector)
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "Failed to get servers: %s\n", err)
-				os.Exit(1)
+				fmt.Fprintf(os.Stderr, "Failed parse Server selector for ServerAuthorization/%s: %s\n", saz.Name, err)
```
Comment (olix0r) [inline pkg/k8s/policy.go:62]:
```suggestion
				fmt.Fprintf(os.Stderr, "Failed to parse Server selector for ServerAuthorization/%s: %s\n", saz.Name, err)
```

File: pkg/k8s/policy.go [hunk 2]
```diff
@@ -55,47 +52,87 @@ func ServerAuthorizationsForResource(ctx context.Context, k8sAPI *KubernetesAPI,
...
+					Server:              server.GetName(),
+					ServerAuthorization: "",
+					AuthorizationPolicy: p.GetName(),
+				})
 			}
+		}
 
-			if serverIncludesPod(server, selectedPods, podSet) {
-				results = append(results, ServerAndAuthorization{server.GetName(), saz.GetName()})
+		if target.Kind == HTTPRouteKind && target.Group == PolicyAPIGroup {
```
Comment (olix0r) [inline pkg/k8s/policy.go:109]: should this be an `else if`? Or should we `continue` in the first block?

File: pkg/k8s/policy.go [hunk 3]
```diff
@@ -55,47 +52,87 @@ func ServerAuthorizationsForResource(ctx context.Context, k8sAPI *KubernetesAPI,
...
+					}
+					if serverIncludesPod(*server, pods) {
+						results = append(results, Authorization{
+							Route:               route.GetName(),
+							Server:              server.GetName(),
+							ServerAuthorization: "",
+							AuthorizationPolicy: p.GetName(),
+						})
+					}
+				}
```
Comment (olix0r) [inline pkg/k8s/policy.go:131]: This block looks identical to the above block. Can it be factored out into its own func?

Comment (adleong) [reply pkg/k8s/policy.go:131]: They are very similar, but yield different Authorization structs and can print different error messages.  We could factor this, but these blocks are so short that we don't end up saving anything, especially considering the number of arguments we would need to pass in.  The `serverIncludesPod` already factors out the majority of the logic.

File: pkg/k8s/policy.go [hunk 4]
```diff
@@ -55,47 +52,85 @@ func ServerAuthorizationsForResource(ctx context.Context, k8sAPI *KubernetesAPI,
...
-			if serverIncludesPod(server, selectedPods, podSet) {
-				results = append(results, ServerAndAuthorization{server.GetName(), saz.GetName()})
+		} else if target.Kind == HTTPRouteKind && target.Group == PolicyAPIGroup {
+			route, err := k8sAPI.L5dCrdClient.PolicyV1alpha1().HTTPRoutes(p.Namespace).Get(ctx, string(target.Name), metav1.GetOptions{})
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "AuthorizationPolicy/%s targets HTTPRoute/%s but we failed get it: %s\n", p.Name, target.Name, err)
+				continue
+			}
+			for _, parent := range route.Spec.ParentRefs {
+				if parent.Kind != nil && *parent.Kind == ServerKind &&
```
Comment (kleimkuhler) [inline pkg/k8s/policy.go:114]: It'd be nice if we add the accessors here for `GetKind()` and `GetGroup`.

Is there a reason we have the `nil` checks for `parent` but not above for `target`?

**Review by kleimkuhler — COMMENTED** [review]
Overall this looks good. Aside from some nits, there are a few comments about the changes to `continue` when we encounter errors, and some expectations around `nil` checks.

File: pkg/k8s/policy.go (hunk 4 above)
Comment (adleong) [reply pkg/k8s/policy.go:114]: ParentRef comes from the gateway-api, so we can't add accessors directly onto it, unfortunately.  ParentRef.Kind is a pointer (and thus needs nil checks) while PolicyTargetRef.Kind is not a pointer.

**Review by kleimkuhler — APPROVED** [review]
👍

**Review by olix0r — APPROVED** [review]

**Pull request merged**